Website/DesktopModules/Admin/Security/Register.ascx.cs

```diff
@@ -726,45 +528,253 @@ private bool Validate()
+
+				if (PortalSettings.Registration.RequireValidProfile)
+				{
+					foreach (ProfilePropertyDefinition property in User.Profile.ProfileProperties)
+					{
+						if (property.Required)
+						{
+							AddProperty(property);
+						}
+					}
+				}
 			}
 			else
 			{
-				if (Request.QueryString["returnurl"] != null)
+				var fields = PortalSettings.Registration.RegistrationFields.Split(',').ToList();
+				//append question/answer field when RequiresQuestionAndAnswer is enabled in config.
+				if (MembershipProviderConfig.RequiresQuestionAndAnswer)
 				{
-					//return to the url passed to register
-					redirectUrl = HttpUtility.UrlDecode(Request.QueryString["returnurl"]);
-					//redirect url should never contain a protocol ( if it does, it is likely a cross-site request forgery attempt )
-					if (redirectUrl.Contains("://") &&
-						!redirectUrl.StartsWith(Globals.AddHTTP(PortalSettings.PortalAlias.HTTPAlias),
-							StringComparison.InvariantCultureIgnoreCase))
+					if (!fields.Contains("PasswordQuestion"))
+					{
+						fields.Add("PasswordQuestion");
+					}
+					if (!fields.Contains("PasswordAnswer"))
 					{
-						redirectUrl = "";
+						fields.Add("PasswordAnswer");
+					}
+				}
+
+				foreach (string field in fields)
+				{
+					var trimmedField = field.Trim();
+					switch (trimmedField)
+					{
+						case "Username":
+							AddField("Username", String.Empty, true, String.IsNullOrEmpty(PortalSettings.Registration.UserNameValidator)
+																		? ExcludeTerms : PortalSettings.Registration.UserNameValidator,
+																		TextBoxMode.SingleLine);
+							break;
+						case "Email":
+							AddField("Email", String.Empty, true, PortalSettings.Registration.EmailValidator, TextBoxMode.SingleLine);
+							break;
+						case "Password":
+							AddPasswordStrengthField(trimmedField, "Membership", true);
+							break;
+						case "PasswordConfirm":
+							AddPasswordConfirmField(trimmedField, "Membership", true);
+							break;
+						case "PasswordQuestion":
+						case "PasswordAnswer":
+							AddField(trimmedField, "Membership", true, String.Empty, TextBoxMode.SingleLine);
+							break;
+						case "DisplayName":
+							AddField(trimmedField, String.Empty, true, ExcludeTerms, TextBoxMode.SingleLine);
+							break;
+						default:
+							ProfilePropertyDefinition property = User.Profile.GetProperty(trimmedField);
+							if (property != null)
+							{
+								AddProperty(property);
+							}
+							break;
 					}
-					if (redirectUrl.Contains("?returnurl"))
+				}
+			}
+
+			//Verify that the current user has access to this page
+			if (PortalSettings.UserRegistration == (int)Globals.PortalRegistrationType.NoRegistration && Request.IsAuthenticated == false)
+			{
+				Response.Redirect(Globals.NavigateURL("Access Denied"), false);
+				Context.ApplicationInstance.CompleteRequest();
+			}
+
+			cancelButton.Click += cancelButton_Click;
```

Comment:
Why was this changed to a button that redirects instead of just a link?  The link that was used before seems more straightforward, is there some reason that the redirect needs to originate on the server?